[toke.c] Small removals of duplicated code

happy-barney opened the pull request:
Improve `toke.c` little bit by removing some duplicated code

Commit:
[toke.c] Move duplicated code into function `S_scan_terminated`

Commit:
[toke.c] Move population of ival into scan_terminated

Commit:
[toke.c] Use 'scan_terminated' in double quote and backtick consistently with single quote

Commit:
[toke.c] Move duplicated code into function `S_warn_expect_operator`

jkeenan commented:
With the caveats that I have never studied the code in `toke.c` before, these
changes basically look good to me.

However, I have one major caveat, which is less a reflection on
@happy-barney's patches than on our QA for `toke.c`.  `toke.c` has many
statements that appear to my eyes to be debugging statements.
```
$ grep -nA1 -E 'printbuf|PerlIO_printf' toke.c
603:        PerlIO_printf(Perl_debug_log, "### %s\n\n", SvPV_nolen_const(report));
604-    };
--
612:S_printbuf(pTHX_ const char *const fmt, const char *const s)
613-{
--
619:    PerlIO_printf(Perl_debug_log, fmt, pv_display(tmp, s, strlen(s), 0, 60));
620-    GCC_DIAG_RESTORE_STMT;
--
2182:        PerlIO_printf(Perl_debug_log, "### forced token:\n");
2183-        tokereport(type, &NEXTVAL_NEXTTOKE);
--
4904:    DEBUG_P(PerlIO_printf(Perl_debug_log, "filter_add func %p (%s)\n",
4905-                          FPTR2DPTR(void *, IoANY(datasv)),
--
4968:    DEBUG_P(PerlIO_printf(Perl_debug_log, "filter_del func %p",
4969-                          FPTR2DPTR(void*, funcp)));
--
5005:        DEBUG_P(PerlIO_printf(Perl_debug_log,
5006-                              "filter_read %d: from rsfp\n", idx));
--
5036:        DEBUG_P(PerlIO_printf(Perl_debug_log,
5037-                              "filter_read %d: skipped (filter deleted)\n",
--
5068:    DEBUG_P(PerlIO_printf(Perl_debug_log,
5069-                          "filter_read %d: via function %p (%s)\n",
--
5727:            PerlIO_printf(Perl_debug_log, "### Saw case modifier\n");
5728-        });
--
5918:            DEBUG_T( { printbuf("### Saw unary minus before =>, forcing word %s\n", s); } );
5919-            OPERATOR(PERLY_MINUS);              /* unary minus */
--
5962:                PerlIO_printf(Perl_debug_log, "### Saw file test %c\n", (int)tmp);
5963-            } );
--
5970:                PerlIO_printf(Perl_debug_log,
5971-                    "### '-%c' looked like a file test but was not\n",
--
6952:    DEBUG_T( { printbuf("### Saw string before %s\n", s); } );
6953-    if (PL_expect == XOPERATOR) {
--
6968:            printbuf("### Saw string before %s\n", s);
6969-        else
6970:            PerlIO_printf(Perl_debug_log,
6971-                         "### Saw unterminated string\n");
--
6998:            printbuf("### Saw backtick string before %s\n", s);
6999-        else
7000:            PerlIO_printf(Perl_debug_log,
7001-                         "### Saw unterminated backtick string\n");
--
9235:                PerlIO_printf(Perl_debug_log, "### Tokener got EOF\n");
9236-            });
--
9585:        DEBUG_T( { printbuf("### Saw number in %s\n", s); } );
9586-        if (PL_expect == XOPERATOR)
--
9757:        PerlIO_printf(Perl_debug_log, "### %" LINE_Tf ":LEX_%s/X%s %s\n",
9758-            CopLINE(PL_curcop),
--
9809:                PerlIO_printf(Perl_debug_log, "### Interpolated variable\n");
9810-        });
--
10038:    DEBUG_T({ PerlIO_printf(Perl_debug_log,
10039-          "### Pending identifier '%s'\n", PL_tokenbuf); });
--
13161:            if (DEBUG_p_TEST || DEBUG_T_TEST) PerlIO_printf(Perl_debug_log, "UTF-16LE script encoding (BOM)\n");
13162-#endif
--
13177:            if (DEBUG_p_TEST || DEBUG_T_TEST) PerlIO_printf(Perl_debug_log, "UTF-16BE script encoding (BOM)\n");
13178-#endif
--
13192:            if (DEBUG_p_TEST || DEBUG_T_TEST) PerlIO_printf(Perl_debug_log, "UTF-8 script encoding (BOM)\n");
13193-#endif
--
13213:                  if (DEBUG_p_TEST || DEBUG_T_TEST) PerlIO_printf(Perl_debug_log, "UTF-16BE script encoding (no BOM)\n");
13214-#endif
--
13231:              if (DEBUG_p_TEST || DEBUG_T_TEST) PerlIO_printf(Perl_debug_log, "UTF-16LE script encoding (no BOM)\n");
13232-#endif
--
13268:    DEBUG_P(PerlIO_printf(Perl_debug_log,
13269-                          "utf16_textfilter(%p,%ce): idx=%d maxlen=%d status=%" IVdf " utf16=%" UVuf " utf8=%" UVuf "\n",
--
13319:                DEBUG_P(PerlIO_printf(Perl_debug_log, "utf16_textfilter partial surrogate detected at %p\n", last_hi));
13320-            }
--
13324:            DEBUG_P(PerlIO_printf(Perl_debug_log, "utf16_textfilter status=%" IVdf " SvCUR(sv)=%" UVuf "\n", status, (UV)SvCUR(utf16_buffer)));
13325-            DEBUG_P({ sv_dump(utf16_buffer); sv_dump(utf8_buffer);});
--
13364:    DEBUG_P(PerlIO_printf(Perl_debug_log,
13365-                          "utf16_textfilter: returns, status=%" IVdf " utf16=%" UVuf " utf8=%" UVuf "\n",
```
Many, perhaps all, of these statements do not appear to be exercised by our
test suite.
```
$ find . -type f | xargs ack -l 'forced token:|filter_add func|filter_del func|filter_read[^:]+: from rsfp|filter_read[^:]+: skipped \(filter deleted\)|filter_read[^:]: via function|Saw case modifier|Saw unary minus before =>, forcing word|Saw file test|looked like a file test but was not|Saw string before|Saw unterminated string|Saw backtick string before|Saw unterminated backtick string|Tokener got EOF|Saw number in|LINE_Tf :LEX_|Interpolated variable|Pending identifier|UTF-16LE script encoding \(BOM\)|UTF-16BE script encoding \(BOM\)|UTF-8 script encoding \(BOM\)|UTF-16BE script encoding \(no BOM\)|UTF-16LE script encoding \(no BOM\)|utf16_textfilter' 
./embed.fnc
./embed.h
./proto.h
dist/Devel-PPPort/parts/embed.fnc
dist/Devel-PPPort/parts/base/5011001
./toke.c
```
The third (especially) and fourth commits in this pull request rework the code
generating these debugging statements.  Without regression tests for such
statements, we run the risk of interfering with existing debugging
functionality.  Should we at the very least consider adding unit tests for the
debugging statements being affected by this p.r. **before** we accept the
p.r.?

happy-barney commented:
@jkeenan debug statements are on list of small refactorings I plan to do.

regarding caveats you mentioned, that's why I also plan to do small PRs (like this one) focusing on 1 or 2 related issues)

jkeenan commented:
> @jkeenan debug statements are on list of small refactorings I plan to do.
> 
> regarding caveats you mentioned, that's why I also plan to do small PRs (like this one) focusing on 1 or 2 related issues)

Okay, but if we don't know *whether* these debug statements are currently being tested, then we should investigate whether they *can* be tested *at all*.  So I would hope you would prioritize those concerns prior to beginning code removal.

happy-barney commented:
I have no idea what you mean by testing debug statements.

tonycoz commented:
> Okay, but if we don't know _whether_ these debug statements are currently being tested, then we should investigate whether they _can_ be tested _at all_. So I would hope you would prioritize those concerns prior to beginning code removal.

In general the exact results of `-D` output isn't something we've documented, and I don't think it's something we guarantee not to change from version to version of perl: there's no deprecation cycle.

I don't think it's something we need to test generally.

We do test some debug output, `re.pm` has some tests for regular expression debug output, and `Devel::Peek` has some tests for sv_dump() output, but both of these are commonly used by end users to debug problems and don't require a `-DDEBUGGING` perl.

tonycoz commented on toke.c:
```diff
@@ -791,6 +791,21 @@ S_scan_terminated(pTHX_ char *s, I32 ival) {
+    if (PL_expect != XOPERATOR)
+        return;
+
+    if (pop_oldbufptr && PL_bufptr > s) {
+        s = PL_bufptr - 1;
+        PL_bufptr = PL_oldbufptr;
+    }
+
```
This could just be moved to S_no_op(), it is static, so not an API, so we can change its signature.

All but one caller to no_op() checks for `XOPERATOR` immediately before the call, and the one exception simply checks a bit earlier.

So I think this check could just be folded into no_op().

tonycoz replied:
It's probably worth renaming no_op() if this is done, I'm not especially fond of no_op_when_xoperator(), maybe check_op_expected(), but other suggestions are welcome.

happy-barney replied:
re: move to `S_no_op` ... done ae77253c5b
re: name ... I'm not very good with names :-) what about `warn_expect_operator` ?

Commit:
[test] Extend `porting/diag.t` to exclude some functions

Excluded functions should not output their parameter(s) via
printf-like format but only as argument(s) of such format.
